source/s3_endpoint.c

```diff
@@ -209,26 +209,27 @@ void aws_s3_endpoint_release(struct aws_s3_endpoint *endpoint) {
         return;
     }
 
-    aws_ref_count_release(&endpoint->ref_count);
-}
-
-void aws_s3_client_endpoint_release(struct aws_s3_client *client, struct aws_s3_endpoint *endpoint) {
-    AWS_PRECONDITION(endpoint);
-    AWS_PRECONDITION(client);
-    AWS_PRECONDITION(endpoint->handled_by_client);
-
-    /* BEGIN CRITICAL SECTION */
-    {
-        aws_s3_client_lock_synced_data(client);
-        /* The last refcount to release */
-        if (aws_atomic_load_int(&endpoint->ref_count.ref_count) == 1) {
-            aws_hash_table_remove(&client->synced_data.endpoints, endpoint->host_name, NULL, NULL);
+    if (endpoint->async_release.task) {
```

Comment:
not a fan of endpoint having 2 "modes", where 1 mode is never used

just needless complexity